File: examples/go/s3_create_bucket.go
// Crea un bucket S3 nuevo en Floci.
// Uso: go run s3_create_bucket.go [nombre-del-bucket]
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

func main() {
	ctx := context.Background()
	cfg, err := config.LoadDefaultConfig(ctx,
		config.WithRegion("us-east-1"),
		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider("test", "test", "")),
	)
	if err != nil {
		log.Fatalf("Error cargando configuracion: %v", err)
	}

	client := s3.NewFromConfig(cfg, func(o *s3.Options) {
		o.UsePathStyle = true
		o.BaseEndpoint = aws.String("http://localhost:4566")
	})

	bucketName := fmt.Sprintf("mi-bucket-%d", time.Now().UnixMilli())
	if len(os.Args) > 1 {
		bucketName = os.Args[1]
	}

	_, err = client.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucketName)})
	if err != nil {
		log.Fatalf("Error creando el bucket (¿ya existe %q?): %v", bucketName, err)
	}
	fmt.Println("Bucket creado:", bucketName)
}

File: examples/rust/s3_create_bucket.rs
// Crea un bucket S3 nuevo en Floci.
// Uso: cargo run --bin s3_create_bucket -- [nombre-del-bucket]

use aws_config::BehaviorVersion;
use aws_credential_types::Credentials;
use aws_sdk_s3::{config::Region, Client, Config};
use std::time::{SystemTime, UNIX_EPOCH};

#[tokio::main]
async fn main() -> Result<(), aws_sdk_s3::Error> {
    let args: Vec<String> = std::env::args().collect();
    let bucket_name = args.get(1).cloned().unwrap_or_else(|| {
        let millis = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis();
        format!("mi-bucket-{millis}")
    });

    let config = Config::builder()
        .behavior_version(BehaviorVersion::latest())
        .endpoint_url("http://localhost:4566")
        .region(Region::new("us-east-1"))
        .credentials_provider(Credentials::new("test", "test", None, None, "floci"))
        .force_path_style(true)
        .build();
    let client = Client::from_conf(config);

    match client.create_bucket().bucket(&bucket_name).send().await {
        Ok(_) => println!("Bucket creado: {bucket_name}"),
        Err(error) => {
            eprintln!("Error creando el bucket (¿ya existe \"{bucket_name}\"?): {error}");
            return Err(error.into());
        }
    }
    Ok(())
}
